internal/app/machined/pkg/controllers/runtime/maintenance_service.go

```diff
@@ -0,0 +1,275 @@
+	return []controller.Input{
+		{
+			Namespace: runtime.NamespaceName,
+			Type:      runtime.MaintenanceServiceRequestType,
+			ID:        pointer.To(runtime.MaintenanceServiceRequestID),
+			Kind:      controller.InputStrong,
+		},
+		{
+			Namespace: runtime.NamespaceName,
+			Type:      runtime.MaintenanceServiceConfigType,
+			ID:        pointer.To(runtime.MaintenanceServiceConfigID),
+			Kind:      controller.InputWeak,
+		},
+		{
+			Namespace: secrets.NamespaceName,
+			Type:      secrets.MaintenanceServiceCertsType,
+			ID:        pointer.To(secrets.MaintenanceServiceCertsID),
+			Kind:      controller.InputWeak,
+		},
+	}
+}
+
+// Outputs implements controller.Controller interface.
+func (ctrl *MaintenanceServiceController) Outputs() []controller.Output {
+	return []controller.Output{
+		{
+			Type: config.MachineConfigType,
+			Kind: controller.OutputShared,
+		},
+	}
+}
+
+// Run implements controller.Controller interface.
+//
+//nolint:gocyclo,cyclop
+func (ctrl *MaintenanceServiceController) Run(ctx context.Context, r controller.Runtime, logger *zap.Logger) error {
+	var (
+		server                     *grpc.Server
+		listener                   net.Listener
+		lastReachableAddresses     []string
+		lastCertificateFingerprint string
+		usagePrinted               bool
+	)
+
+	shutdownServer := func(ctx context.Context) {
+		if server != nil {
+			shutdownCtx, shutdownCancel := context.WithTimeout(ctx, 5*time.Second)
+			defer shutdownCancel()
+
+			factory.ServerGracefulStop(server, shutdownCtx)
+
+			server = nil
+		}
+
+		if listener != nil {
+			listener.Close() //nolint:errcheck
+
+			listener = nil
+			lastReachableAddresses = nil
+		}
+	}
+
+	defer shutdownServer(context.Background())
+
+	cfgCh := make(chan machineryconfig.Provider)
+	srv := maintenance.New(cfgCh)
+	injector := &authz.Injector{
+		Mode:   authz.ReadOnly,
+		Logger: logger.Sugar().Debugf,
+	}
+	tlsProvider := maintenance.NewTLSProvider()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		case cfg := <-cfgCh:
+			configResource := config.NewMachineConfigWithID(cfg, config.MaintenanceID)
+
+			oldConfigResource, err := safe.ReaderGetByID[*config.MachineConfig](ctx, r, config.MaintenanceID)
+			if err != nil && !state.IsNotFoundError(err) {
+				return fmt.Errorf("failed to get machine config: %w", err)
+			}
+
+			if state.IsNotFoundError(err) {
+				if err = r.Create(ctx, configResource); err != nil {
+					return fmt.Errorf("failed to create machine config: %w", err)
+				}
+			} else {
+				configResource.Metadata().SetVersion(oldConfigResource.Metadata().Version())
+
+				if err = configResource.Metadata().SetOwner(oldConfigResource.Metadata().Owner()); err != nil {
+					return fmt.Errorf("error setting owner: %w", err)
+				}
+
+				if err = r.Update(ctx, configResource); err != nil {
+					return fmt.Errorf("failed to update machine config: %w", err)
+				}
+			}
+
+			continue
```

Comment:
Can we replace all that with `safe.WriterModify(...)`?

Reply:
this resource is a bit special, it doesn't support changing the spec, and WriterModify requires a "null" value which is not really defined for machine config